File: macos/VusbClient/Network/VusbNetworkClient.swift
//
//  VusbNetworkClient.swift
//  VusbClient
//
//  Network client for VUSB protocol communication
//

import Foundation
import Network
import Combine
import os.log

/// Connection state for the VUSB client
enum ConnectionState: Equatable {
    case disconnected
    case connecting
    case connected(serverAddress: String)
    case error(message: String)
    
    var isConnected: Bool {
        if case .connected = self { return true }
        return false
    }
}

/// Network client for VUSB protocol
@MainActor
class VusbNetworkClient: ObservableObject {
    
    // MARK: - Constants
    
    private static let connectTimeout: TimeInterval = 10.0
    private static let readTimeout: TimeInterval = 30.0
    private static let keepAliveInterval: TimeInterval = 10.0
    
    private let logger = Logger(subsystem: "com.vusb.client", category: "Network")
    
    // MARK: - Published Properties
    
    @Published private(set) var connectionState: ConnectionState = .disconnected
    @Published private(set) var lastError: String?
    
    // MARK: - Private Properties
    
    private var connection: NWConnection?
    private var sequenceNumber: UInt32 = 0
    private var isRunning = false
    
    private var keepAliveTimer: Timer?
    private var pendingRequests: [UInt32: CheckedContinuation<(VusbHeader, Data), Error>] = [:]
    
    // Callbacks
    var onMessageReceived: ((VusbHeader, Data) -> Void)?
    var onUrbSubmit: ((UrbSubmit) -> UrbComplete?)?
    var onDisconnected: (() -> Void)?
    
    // MARK: - Connection Management
    
    /// Connect to VUSB server
    func connect(serverAddress: String, port: UInt16 = VUSB_DEFAULT_PORT, clientName: String = "macOSClient") async -> Bool {
        guard !isRunning else {
            logger.warning("Already connected or connecting")
            return false
        }
        
        connectionState = .connecting
        
        let host = NWEndpoint.Host(serverAddress)
        let port = NWEndpoint.Port(rawValue: port)!
        
        let parameters = NWParameters.tcp
        parameters.allowLocalEndpointReuse = true
        
        connection = NWConnection(host: host, port: port, using: parameters)
        
        return await withCheckedContinuation { continuation in
            connection?.stateUpdateHandler = { [weak self] state in
                guard let self = self else { return }
                
                Task { @MainActor in
                    switch state {
                    case .ready:
                        self.logger.info("Connected to \(serverAddress):\(port.rawValue)")
                        self.isRunning = true
                        
                        // Send connect message
                        do {
                            let connectMsg = ConnectMessage(clientName: clientName)
                            try await self.sendMessage(command: .connect, payload: connectMsg.toData())
                            
                            // Wait for connect response (server echoes CONNECT command with status)
                            let (header, _) = try await self.receiveMessage()
                            
                            if header.commandType == .connect {
                                self.connectionState = .connected(serverAddress: serverAddress)
                                self.startReceiveLoop()
                                self.startKeepAlive()
                                continuation.resume(returning: true)
                            } else {
                                throw NSError(domain: "VusbClient", code: -1, 
                                            userInfo: [NSLocalizedDescriptionKey: "Expected CONNECT response"])
                            }
                        } catch {
                            self.logger.error("Connection handshake failed: \(error.localizedDescription)")
                            self.connectionState = .error(message: error.localizedDescription)
                            self.disconnect()
                            continuation.resume(returning: false)
                        }
                        
                    case .failed(let error):
                        self.logger.error("Connection failed: \(error.localizedDescription)")
                        self.connectionState = .error(message: error.localizedDescription)
                        self.isRunning = false
                        continuation.resume(returning: false)
                        
                    case .cancelled:
                        self.logger.info("Connection cancelled")
                        self.connectionState = .disconnected
                        self.isRunning = false
                        
                    case .waiting(let error):
                        self.logger.warning("Connection waiting: \(error.localizedDescription)")
                        
                    default:
                        break
                    }
                }
            }
            
            connection?.start(queue: .global(qos: .userInitiated))
        }
    }
    
    /// Disconnect from server
    func disconnect() {
        guard isRunning else { return }
        
        logger.info("Disconnecting...")
        isRunning = false
        
        // Stop keep-alive timer
        keepAliveTimer?.invalidate()
        keepAliveTimer = nil
        
        // Send disconnect message (best effort)
        Task {
            try? await sendMessage(command: .disconnect, payload: Data())
        }
        
        // Cancel connection
        connection?.cancel()
        connection = nil
        
        // Clear pending requests
        for (_, continuation) in pendingRequests {
            continuation.resume(throwing: NSError(domain: "VusbClient", code: -2,
                                                  userInfo: [NSLocalizedDescriptionKey: "Disconnected"]))
        }
        pendingRequests.removeAll()
        
        connectionState = .disconnected
        onDisconnected?()
        
        logger.info("Disconnected")
    }
    
    // MARK: - Message Sending
    
    /// Send a protocol message
    func sendMessage(command: VusbCommand, payload: Data) async throws {
        guard let connection = connection, isRunning else {
            throw NSError(domain: "VusbClient", code: -3,
                         userInfo: [NSLocalizedDescriptionKey: "Not connected"])
        }
        
        sequenceNumber += 1
        let header = VusbHeader(command: command, length: UInt32(payload.count), sequence: sequenceNumber)
        
        var messageData = header.toData()
        messageData.append(payload)
        
        logger.debug("Sending message: cmd=0x\(String(format: "%04X", command.rawValue)), len=\(payload.count), seq=\(self.sequenceNumber)")
        
        return try await withCheckedThrowingContinuation { continuation in
            connection.send(content: messageData, completion: .contentProcessed { error in
                if let error = error {
                    continuation.resume(throwing: error)
                } else {
                    continuation.resume()
                }
            })
        }
    }
    
    /// Send message and wait for response
    func sendMessageWithResponse(command: VusbCommand, payload: Data) async throws -> (VusbHeader, Data) {
        try await sendMessage(command: command, payload: payload)
        return try await receiveMessage()
    }
    
    // MARK: - Message Receiving
    
    /// Receive a single message
    private func receiveMessage() async throws -> (VusbHeader, Data) {
        guard let connection = connection else {
            throw NSError(domain: "VusbClient", code: -3,
                         userInfo: [NSLocalizedDescriptionKey: "Not connected"])
        }
        
        // Receive header
        let headerData = try await receiveExact(connection: connection, length: VUSB_HEADER_SIZE)
        
        guard let header = VusbHeader.fromData(headerData) else {
            throw NSError(domain: "VusbClient", code: -4,
                         userInfo: [NSLocalizedDescriptionKey: "Invalid header"])
        }
        
        // Validate magic
        guard header.magic == VUSB_PROTOCOL_MAGIC else {
            throw NSError(domain: "VusbClient", code: -5,
                         userInfo: [NSLocalizedDescriptionKey: "Invalid magic number"])
        }
        
        // Receive payload
        var payload = Data()
        if header.length > 0 {
            payload = try await receiveExact(connection: connection, length: Int(header.length))
        }
        
        logger.debug("Received message: cmd=0x\(String(format: "%04X", header.command)), len=\(header.length), seq=\(header.sequence)")
        
        return (header, payload)
    }
    
    /// Receive exact number of bytes
    private func receiveExact(connection: NWConnection, length: Int) async throws -> Data {
        return try await withCheckedThrowingContinuation { continuation in
            connection.receive(minimumIncompleteLength: length, maximumLength: length) { content, _, isComplete, error in
                if let error = error {
                    continuation.resume(throwing: error)
                } else if let data = content {
                    continuation.resume(returning: data)
                } else if isComplete {
                    continuation.resume(throwing: NSError(domain: "VusbClient", code: -6,
                                                         userInfo: [NSLocalizedDescriptionKey: "Connection closed"]))
                } else {
                    continuation.resume(returning: Data())
                }
            }
        }
    }
    
    // MARK: - Background Tasks
    
    /// Start the receive loop for incoming messages
    private func startReceiveLoop() {
        Task.detached { [weak self] in
            while let self = self, await self.isRunning {
                do {
                    let (header, payload) = try await self.receiveMessage()
                    await self.handleMessage(header: header, payload: payload)
                } catch {
                    await MainActor.run {
                        if self.isRunning {
                            self.logger.error("Receive error: \(error.localizedDescription)")
                            self.disconnect()
                        }
                    }
                    break
                }
            }
        }
    }
    
    /// Handle received message
    @MainActor
    private func handleMessage(header: VusbHeader, payload: Data) {
        switch header.commandType {
        case .ping:
            // Respond with pong
            Task {
                try? await sendMessage(command: .pong, payload: Data())
            }
            
        case .pong:
            // Response to our ping - connection is alive
            logger.debug("Received pong from server")
            break
            
        case .urbSubmit:
            // Handle URB request from server
            if let urb = UrbSubmit.fromData(payload) {
                if let response = onUrbSubmit?(urb) {
                    Task {
                        try? await sendMessage(command: .urbComplete, payload: response.toData())
                    }
                }
            }
            
        case .error:
            // Handle error message
            if let errorMessage = String(data: payload, encoding: .utf8) {
                logger.error("Server error: \(errorMessage)")
                lastError = errorMessage
            }
            
        default:
            // Pass to general handler
            onMessageReceived?(header, payload)
        }
    }
    
    /// Start keep-alive timer
    private func startKeepAlive() {
        keepAliveTimer = Timer.scheduledTimer(withTimeInterval: Self.keepAliveInterval, repeats: true) { [weak self] _ in
            guard let self = self else { return }
            Task {
                try? await self.sendMessage(command: .ping, payload: Data())
            }
        }
    }
    
    // MARK: - Device Operations
    
    /// Attach a device to the server
    /// Sends VUSB_DEVICE_INFO (208 bytes) + descriptorLength (4 bytes) + descriptors
    func attachDevice(_ device: VusbDeviceInfo) async throws {
        var payload = device.toData()
        
        // Combine device and config descriptors
        var descriptors = Data()
        descriptors.append(device.deviceDescriptor)
        descriptors.append(device.configDescriptor)
        
        // Append descriptor length (4 bytes) and descriptor data
        payload.append(contentsOf: withUnsafeBytes(of: UInt32(descriptors.count).littleEndian) { Array($0) })
        payload.append(descriptors)
        
        try await sendMessage(command: .deviceAttach, payload: payload)
        logger.info("Device attached: \(device.displayName)")
    }
    
    /// Detach a device from the server
    func detachDevice(deviceId: Int32) async throws {
        var data = Data()
        data.append(contentsOf: withUnsafeBytes(of: deviceId.littleEndian) { Array($0) })
        try await sendMessage(command: .deviceDetach, payload: data)
        logger.info("Device detached: \(deviceId)")
    }
    
    /// Send URB completion
    func sendUrbComplete(_ completion: UrbComplete) async throws {
        try await sendMessage(command: .urbComplete, payload: completion.toData())
    }
}

// MARK: - Async extension for isRunning check

extension VusbNetworkClient {
    nonisolated var isRunning: Bool {
        get async {
            await MainActor.run { self.isRunning }
        }
    }
    
    private func setIsRunning(_ value: Bool) {
        self.isRunning = value
    }
}

File: android/app/src/main/java/com/vusb/client/network/VusbNetworkClient.kt
package com.vusb.client.network

import android.util.Log
import com.vusb.client.protocol.*
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import java.io.DataInputStream
import java.io.DataOutputStream
import java.net.InetSocketAddress
import java.net.Socket
import java.net.SocketTimeoutException
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicInteger

/**
 * Network client for VUSB protocol
 * 
 * Handles:
 * - TCP connection to server
 * - Protocol message encoding/decoding
 * - Message sending and receiving
 * - Keep-alive mechanism
 */
class VusbNetworkClient {
    
    companion object {
        private const val TAG = "VusbNetworkClient"
        private const val CONNECT_TIMEOUT = 10000 // 10 seconds
        private const val READ_TIMEOUT = 30000 // 30 seconds
        private const val KEEPALIVE_INTERVAL = 10000L // 10 seconds
    }
    
    // Connection state
    sealed class ConnectionState {
        data object Disconnected : ConnectionState()
        data object Connecting : ConnectionState()
        data class Connected(val serverAddress: String) : ConnectionState()
        data class Error(val message: String) : ConnectionState()
    }
    
    private var socket: Socket? = null
    private var outputStream: DataOutputStream? = null
    private var inputStream: DataInputStream? = null
    
    private val sequenceNumber = AtomicInteger(0)
    private val isRunning = AtomicBoolean(false)
    
    private val _connectionState = MutableStateFlow<ConnectionState>(ConnectionState.Disconnected)
    val connectionState: StateFlow<ConnectionState> = _connectionState.asStateFlow()
    
    private var scope: CoroutineScope? = null
    private var receiveJob: Job? = null
    private var keepAliveJob: Job? = null
    
    // Message callback
    var onMessageReceived: ((VusbHeader, ByteArray) -> Unit)? = null
    var onDisconnected: (() -> Unit)? = null
    
    /**
     * Connect to the VUSB server
     */
    suspend fun connect(
        serverAddress: String, 
        port: Int = VusbProtocol.DEFAULT_PORT,
        clientName: String = "AndroidClient"
    ): Boolean = withContext(Dispatchers.IO) {
        
        if (isRunning.get()) {
            Log.w(TAG, "Already connected or connecting")
            return@withContext false
        }
        
        _connectionState.value = ConnectionState.Connecting
        
        try {
            Log.d(TAG, "Connecting to $serverAddress:$port")
            
            socket = Socket().apply {
                soTimeout = READ_TIMEOUT
                connect(InetSocketAddress(serverAddress, port), CONNECT_TIMEOUT)
            }
            
            outputStream = DataOutputStream(socket!!.getOutputStream())
            inputStream = DataInputStream(socket!!.getInputStream())
            
            isRunning.set(true)
            
            // Send connect message
            val connectMsg = ConnectMessage(clientName)
            sendMessage(VusbProtocol.Command.CONNECT, connectMsg.toByteArray())
            
            // Wait for connect response (server echoes CONNECT command with status)
            val (header, _) = receiveMessage()
            if (header.command != VusbProtocol.Command.CONNECT) {
                throw Exception("Expected CONNECT response, got ${header.command}")
            }
            
            Log.d(TAG, "Connected successfully to $serverAddress:$port")
            _connectionState.value = ConnectionState.Connected(serverAddress)
            
            // Start background tasks
            scope = CoroutineScope(Dispatchers.IO + SupervisorJob())
            startReceiveLoop()
            startKeepAlive()
            
            true
        } catch (e: Exception) {
            Log.e(TAG, "Connection failed", e)
            _connectionState.value = ConnectionState.Error(e.message ?: "Connection failed")
            disconnect()
            false
        }
    }
    
    /**
     * Disconnect from server
     */
    fun disconnect() {
        if (!isRunning.getAndSet(false)) {
            return
        }
        
        Log.d(TAG, "Disconnecting...")
        
        // Cancel background jobs
        receiveJob?.cancel()
        keepAliveJob?.cancel()
        scope?.cancel()
        
        // Send disconnect message (best effort)
        try {
            sendMessageSync(VusbProtocol.Command.DISCONNECT, ByteArray(0))
        } catch (e: Exception) {
            Log.w(TAG, "Failed to send disconnect message", e)
        }
        
        // Close streams and socket
        try {
            outputStream?.close()
            inputStream?.close()
            socket?.close()
        } catch (e: Exception) {
            Log.w(TAG, "Error closing connection", e)
        }
        
        outputStream = null
        inputStream = null
        socket = null
        
        _connectionState.value = ConnectionState.Disconnected
        onDisconnected?.invoke()
        
        Log.d(TAG, "Disconnected")
    }
    
    /**
     * Check if connected
     */
    fun isConnected(): Boolean {
        return isRunning.get() && socket?.isConnected == true && !socket!!.isClosed
    }
    
    /**
     * Send a protocol message
     */
    suspend fun sendMessage(command: Short, payload: ByteArray) = withContext(Dispatchers.IO) {
        sendMessageSync(command, payload)
    }
    
    /**
     * Send message synchronously (for use in IO context)
     */
    @Synchronized
    private fun sendMessageSync(command: Short, payload: ByteArray) {
        val output = outputStream ?: throw Exception("Not connected")
        
        val header = VusbHeader(
            command = command,
            length = payload.size,
            sequence = sequenceNumber.incrementAndGet()
        )
        
        val headerBytes = header.toByteArray()
        
        Log.v(TAG, "Sending message: cmd=0x${command.toString(16)}, len=${payload.size}, seq=${header.sequence}")
        
        output.write(headerBytes)
        if (payload.isNotEmpty()) {
            output.write(payload)
        }
        output.flush()
    }
    
    /**
     * Receive a single message
     */
    @Synchronized
    private fun receiveMessage(): Pair<VusbHeader, ByteArray> {
        val input = inputStream ?: throw Exception("Not connected")
        
        // Read header
        val headerBytes = ByteArray(VusbProtocol.HEADER_SIZE)
        input.readFully(headerBytes)
        
        val header = VusbHeader.fromByteArray(headerBytes)
        
        // Verify magic
        if (header.magic != VusbProtocol.MAGIC) {
            throw Exception("Invalid protocol magic: 0x${header.magic.toString(16)}")
        }
        
        // Read payload
        val payload = if (header.length > 0) {
            if (header.length > VusbProtocol.MAX_PAYLOAD_SIZE) {
                throw Exception("Payload too large: ${header.length}")
            }
            ByteArray(header.length).also { input.readFully(it) }
        } else {
            ByteArray(0)
        }
        
        Log.v(TAG, "Received message: cmd=0x${header.command.toString(16)}, len=${payload.size}, seq=${header.sequence}")
        
        return Pair(header, payload)
    }
    
    /**
     * Start the receive loop
     */
    private fun startReceiveLoop() {
        receiveJob = scope?.launch {
            while (isActive && isRunning.get()) {
                try {
                    val (header, payload) = receiveMessage()
                    
                    when (header.command) {
                        VusbProtocol.Command.PONG -> {
                            // Response to our ping
                            Log.v(TAG, "Received pong")
                        }
                        VusbProtocol.Command.DISCONNECT -> {
                            Log.d(TAG, "Server disconnected")
                            disconnect()
                            break
                        }
                        else -> {
                            // Dispatch to message handler
                            onMessageReceived?.invoke(header, payload)
                        }
                    }
                } catch (e: SocketTimeoutException) {
                    // Timeout is expected, continue
                    continue
                } catch (e: Exception) {
                    if (isRunning.get()) {
                        Log.e(TAG, "Receive error", e)
                        _connectionState.value = ConnectionState.Error(e.message ?: "Receive error")
                        disconnect()
                    }
                    break
                }
            }
        }
    }
    
    /**
     * Start keep-alive mechanism
     */
    private fun startKeepAlive() {
        keepAliveJob = scope?.launch {
            while (isActive && isRunning.get()) {
                delay(KEEPALIVE_INTERVAL)
                try {
                    if (isRunning.get()) {
                        sendMessage(VusbProtocol.Command.PING, ByteArray(0))
                    }
                } catch (e: Exception) {
                    Log.w(TAG, "Keepalive failed", e)
                }
            }
        }
    }
    
    /**
     * Attach a device to the server
     */
    suspend fun attachDevice(deviceInfo: DeviceInfo) {
        Log.d(TAG, "Attaching device: VID=${deviceInfo.vendorId.toString(16)}, PID=${deviceInfo.productId.toString(16)}")
        sendMessage(VusbProtocol.Command.ATTACH, deviceInfo.toByteArray())
    }
    
    /**
     * Detach a device from the server
     */
    suspend fun detachDevice(deviceId: Int) {
        Log.d(TAG, "Detaching device: $deviceId")
        val buffer = ByteBuffer.allocate(4)
        buffer.order(ByteOrder.LITTLE_ENDIAN)
        buffer.putInt(deviceId)
        sendMessage(VusbProtocol.Command.DETACH, buffer.array())
    }
    
    /**
     * Send URB completion response
     */
    suspend fun sendUrbComplete(urbComplete: UrbComplete) {
        Log.d(TAG, "Sending URB complete: id=${urbComplete.urbId}, status=${urbComplete.status}, len=${urbComplete.actualLength}")
        sendMessage(VusbProtocol.Command.URB_COMPLETE, urbComplete.toByteArray())
    }
}
